File: backend/src/models/Member.ts
import mongoose, { Schema } from "mongoose";

const MemberSchema = new Schema(
  {
    // Unique numeric ID handed to gym staff so they can key it into the
    // Hikvision terminal's own menu while enrolling this member's
    // face/fingerprint (employeeNo on the device).
    deviceUserId: { type: String, required: true, unique: true },

    name: { type: String, required: true },
    phone: { type: String, required: true },
    email: { type: String },
    gender: { type: String, enum: ["male", "female", "other"] },
    photoUrl: { type: String },
    address: { type: String },
    emergencyContact: { type: String },
    joinDate: { type: Date, default: Date.now },

    currentPlan: { type: Schema.Types.ObjectId, ref: "MembershipPlan" },
    membershipStart: { type: Date },
    membershipEnd: { type: Date },
    status: {
      type: String,
      enum: ["active", "expired", "inactive"],
      default: "inactive",
    },

    biometricEnrolled: { type: Boolean, default: false },
    device: { type: Schema.Types.ObjectId, ref: "Device" },

    notes: { type: String },
  },
  { timestamps: true }
);

export default mongoose.models.Member || mongoose.model("Member", MemberSchema);

File: mobile/lib/models/member.dart
import 'plan.dart';

class Member {
  final String id;
  final String deviceUserId;
  final String name;
  final String phone;
  final String? email;
  final String? gender;
  final String? address;
  final DateTime? membershipStart;
  final DateTime? membershipEnd;
  final String status; // active | expired | inactive
  final bool biometricEnrolled;
  final MembershipPlan? currentPlan;

  Member({
    required this.id,
    required this.deviceUserId,
    required this.name,
    required this.phone,
    this.email,
    this.gender,
    this.address,
    this.membershipStart,
    this.membershipEnd,
    required this.status,
    required this.biometricEnrolled,
    this.currentPlan,
  });

  factory Member.fromJson(Map<String, dynamic> json) {
    return Member(
      id: json['_id'],
      deviceUserId: json['deviceUserId'] ?? '',
      name: json['name'] ?? '',
      phone: json['phone'] ?? '',
      email: json['email'],
      gender: json['gender'],
      address: json['address'],
      membershipStart: json['membershipStart'] != null
          ? DateTime.parse(json['membershipStart'])
          : null,
      membershipEnd: json['membershipEnd'] != null
          ? DateTime.parse(json['membershipEnd'])
          : null,
      status: json['status'] ?? 'inactive',
      biometricEnrolled: json['biometricEnrolled'] ?? false,
      currentPlan: json['currentPlan'] is Map<String, dynamic>
          ? MembershipPlan.fromJson(json['currentPlan'])
          : null,
    );
  }

  int get daysLeft {
    if (membershipEnd == null) return 0;
    return membershipEnd!.difference(DateTime.now()).inDays;
  }
}
